packages/@aws-cdk/aws-redshift/lib/parameter-group.ts

```diff
@@ -12,6 +12,19 @@ export interface IClusterParameterGroup extends IResource {
    * @attribute
    */
   readonly clusterParameterGroupName: string;
+
+  /**
+   * Adds a parameter to the parameter group
+   *
+   * @param name the parameter name
+   * @param value the parameter name
+   * @returns metadata about the execution of this method. If the parameter
+   * was not added, the value of `parameterAddedResult` will report a failure status. You
+   * should always check this value to make sure that the operation was
+   * actually carried out. Otherwise, synthesis and deploy will terminate
+   * silently, which may be confusing.
```

Comment:
I pretty fundamentally disagree with this approach. You should throw an error if it fails. Requiring an extra success check is not a good user experience.

Reply:
I agree that this isn't a great user experience,  but the different failure states are there so that it can be handled appropriately.

For example on `SAME_VALUE_FAILURE` the parameter technically didn't get added, but already exists in the desired state

`CONFLICTING_VALUE_FAILURE` would tell the user that the same parameter exists but was given a different value elsewhere (would likely throw an error here)

`IMPORTED_RESOURCE_FAILURE` would tell a user that they can't add to an imported parameter group or cluster.

I was going to originally have continue execution on a successfully addition or  if the parameter already exists in the desired state (even though it didn't get added)  and throw an error for  a conflicting value and imported resources. I think I may be too hung up on the fact that the parameter doesn't technically get added if it already exists in the desired state

Reply:
It's definitely the right call to have descriptive error messages / error codes. However, all of this information is better expressed through actual errors, because if the failure occurs, then execution should stop. If it isn't a real failure (eg they're adding something that already exists) then it's fine to continue silently, because that isn't actually a failure.